#!/bin/bash
set -e 
#######################################
# load environment variables for LRSDAY
source ./../../env.sh

#######################################
# set project-specific variables
prefix="SK1" # The file name prefix for the processing sample. Default = "SK1" for the testing example.
genome="./../07.Supervised_Final_Assembly/$prefix.assembly.final.fa" # The file path of the input genome assembly.
centromere_gff3="./../08.Centromere_Annotation/$prefix.centromere.gff3" # The path of the final centromere annotation gff3 generated by task 08.Centromere_Annotation_for_FinalAssembly, use "" if not existing.
TE_gff3="./../11.TE_Annotation/$prefix.TE.gff3" # The file path of the TE annotation gff3 generated by task 11.TE_Annotation, use "" if not existing.
X_element_gff3="./../12.Core_X_Element_Annotation/$prefix.X_element.gff3" # The file path of the X_element annotation gff3 generated by task 12.CoreX_Annotation, use "" if not existing.
Y_prime_element_gff3="./../13.Y_Prime_Element_Annotation/$prefix.Y_prime_element.gff3" # The file path of the Y_prime_element annotation gff3 generated by task 13.Yprime_Annotation, use "" if not existing.
nuclear_gene_gff3="./../14.Gene_Orthology_Identification/$prefix.nuclear_gene.updated.gff3" # The file path of the nuclear gene annotation gff3 generated by task 14.GeneOrthology_Identification (which is better) or 09.Nuclear_Gene_Annotation, use "" if not existing.
mitochondrial_gene_gff3="./../14.Gene_Orthology_Identification/$prefix.mitochondrial_gene.updated.gff3" # The file path of the mitochondrial gene annotation gff3 generated by task 14.GeneOrthology_Identification (which is better) or 10.Mitochondrial_Gene_Annotation, use "" if not existing.
chrMT_tag="chrMT" # The sequence name for the mitochondrial genome in the input genome assembly, if there are multiple corresponding contigs/scaffolds, use a single ';' to separate them. e.g. "chrMT_1;chrMT_2". Default = "chrMT". 
chrMT_genetic_code_table=3 # The NCBI genetic code table (https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi) for the annotated mitochondrial genome. Default = 3 (i.e. Yeast Mitochondria)

debug="no" # use "yes" if prefer to keep intermediate files, otherwise use "no".

######################################
# process the pipeline
cat $centromere_gff3 $TE_gff3 $X_element_gff3 $Y_prime_element_gff3 $nuclear_gene_gff3 $mitochondrial_gene_gff3  >$prefix.concatenated.gff3
perl $LRSDAY_HOME/scripts/sort_gff3.pl -i $prefix.concatenated.gff3 -t $prefix -o $prefix.final.gff3 -r $genome
perl $LRSDAY_HOME/scripts/extract_cds_from_tidy_gff3.pl -r $genome -g $prefix.final.gff3 -o $prefix.final.cds.fa

cp $genome $prefix.assembly.final.fa
echo $chrMT_tag | sed -e "s/;/\n/g" > $prefix.assembly.chrMT.list
$LRSDAY_HOME/scripts/select_fasta_by_list.pl -i $genome -l $prefix.assembly.chrMT.list -m reverse -o $prefix.assembly.nuclear_genome.fa
$LRSDAY_HOME/scripts/select_fasta_by_list.pl -i $genome -l $prefix.assembly.chrMT.list -m normal -o $prefix.assembly.mitochondrial_genome.fa

if [[ ! -z $nuclear_gene_gff3 ]]
then
    perl $LRSDAY_HOME/scripts/extract_cds_from_tidy_gff3.pl -r $prefix.assembly.nuclear_genome.fa -g $prefix.final.gff3 -o $prefix.nuclear_genome.cds.fa
    perl $LRSDAY_HOME/scripts/cds2protein.pl -i $prefix.nuclear_genome.cds.fa -p $prefix.nuclear_genome -t 1
fi

if [[ ! -z $mitochondrial_gene_gff3 ]]
then
    perl $LRSDAY_HOME/scripts/extract_cds_from_tidy_gff3.pl -r $prefix.assembly.mitochondrial_genome.fa -g $prefix.final.gff3 -o $prefix.mitochondrial_genome.cds.fa
    perl $LRSDAY_HOME/scripts/cds2protein.pl -i $prefix.mitochondrial_genome.cds.fa -p $prefix.mitochondrial_genome -t $chrMT_genetic_code_table
fi

if [[ ! -z $nuclear_gene_gff3 && ! -z $mitochondrial_gene_gff3 ]]
then
    cat $prefix.nuclear_genome.trimmed_cds.fa $prefix.mitochondrial_genome.trimmed_cds.fa > $prefix.final.trimmed_cds.fa
    cat $prefix.nuclear_genome.pep.fa $prefix.mitochondrial_genome.pep.fa >$prefix.final.pep.fa
    cat $prefix.nuclear_genome.manual_check.list  $prefix.mitochondrial_genome.manual_check.list > $prefix.final.manual_check.list
fi

# clean up intermediate files
if [[ $debug == "no" ]]
then
    rm $prefix.concatenated.gff3
fi

############################
# checking bash exit status
if [[ $? -eq 0 ]]
then
    echo ""
    echo "LRSDAY message: This bash script has been successfully processed! :)"
    echo ""
    echo ""
    exit 0
fi
############################
